index.bs

```diff
@@ -679,11 +678,87 @@ are traditionally added to {{WindowEventHandlers}} rather than {{Window}}.</p>
+If the complexity of the algorithm is reasonably understood, you may add appropriate
+state to track that the algorithm is in progress and terminate early if the algorithm
+is already concurrently running. This technique is "guarding" the algorithm:
+
+<div class="example">
+    A technique (e.g., used for [abort()](https://dom.spec.whatwg.org/#abortsignal-add)
+    or as a stack-based variation in
+    [focus()](https://html.spec.whatwg.org/multipage/interaction.html#focus-update-steps))
+    for guarding against unplanned recursion. This example is given in JavaScript
+    for terseness but would be applied in algorithmic prose in a spec.
+
+    <pre highlight="js">
+    const isStarted = false;
+
+    function doComplexOperationWithEvents() {
+        if ( isStarted ) {
+            return; // Terminate early
+        }
+        isStarted = true;
+        // Do complex stuff
+        
+        // Notify the developer of progress by firing events
+        //   Note: any other code may run here, including calling 
+        //   doComplexOperationWithEvents() again!
+
+        // Finishe doing complex stuff
+        isStarted = false;        
+    }
+    </pre>
+</div>
+
+Sometimes "guarding" as illustrated above cannot be done because of the complexity
+of the algorithm, the number of potential algorithm entrypoints or the side-effects
+beyond the algorithm's control. In these cases another form of protecting the integrity
+of the algorithm's state is to "defer" the event to a subsequent task or microtask.
+Deferral ensures that any stack-based recursion is avoided (but does not eliminate
+potentially problematic loops, as they could now occur as unending follow-up tasks).
+
+Defering an event is often specified as "[=queue a task=] to [=fire an event=]...".
+
+By way of illustration, the deprecated
+[Mutation Events](https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Mutation_events),
+were specified to fire in the middle of node insertion and removal algorithms, which
+are triggered by a variety of common API calls like {{Node/appendChild()}}. The side-effects
+from recursivly running node mutation algorithms caused many years of security issues.
+A more modern approach embodied in
+[Mutation Observers](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver)
+applies the principle of deferral of notification--in this case to the microtask queue
+following the algorithm.
+
+Weigh the trade-offs when considering deferral as an alternative to guarding as noted
+earlier. Guarding is generally the preferred approach. With guarding:
+
+* all state management can be rationalized within one "turn" of the algorithm (no
+    need to consider any state changes or API invocations between the time the algorithm
+    completes and when the event fires).
+* events can reveal internal state changes as early as possible when dispatched immediately
```

Comment:
This is exactly it. Between `conditionsAreRight` and the time that `queueATask` fires, conditions may have changed (since who knows that task will have been picked up and run before your queued task). This appears to be annevk's concern with recommending queuing a task as the default suggestion back in issue #81.

Reply:
True, something could invalidate the task queue... like maybe the document becoming inactive. It's a hard problem. I'm not sure what we can suggest here apart from... here are a bunch of things to think hard about...

Reply:
Hopefully the list of things to think about will be helpful. Agree that it's not strictly cut and dry.